### PR TITLE
Save report to a file

### LINKED ISSUE #17: Save report to a file, text and JSON formats
Please allow to save the report in a file, in either text format, as now outputted, and as a JSON.

Perhaps:

```js
reportFormat: 'text',
reportFile: 'report-file.txt'
```
and

```js
reportFormat: 'json',
reportFile: 'report-file.json'
```

### PR DESCRIPTION
**Pull request by borisdiakur**
This pull request resolves #17 <strike>by hooking into `process.stdout` and writing the reporter output to a file</strike>. To do that the user can set one additional option `outputPath`. The path includes filename and extention (two options, reportFormat and reportFile, seemed redundant).
I also did some cleanup, hope you don't mind (see commits).

### CODE REVIEW

**Comment by stefanjudis** [comment]
@borisdiakur 

Oh wow - that's what I call a PR. Nice one!

Will have a look right now. :) Thanks.

File: test/jsinspect_test.js [hunk 1]
```diff
@@ -249,6 +249,31 @@ exports.jsinspect = {
           }
         );
       }
+    },
+
+    reportToFile: function(test) {
+      fs.mkdir('tmp', function () {
```
Comment (stefanjudis) [inline test/jsinspect_test.js]: You're creating a directory here, which won't be removed after test run.
We may want to remove it afterwards?

File: README.md [hunk 2]
```diff
@@ -112,6 +112,19 @@ Specify the reporter to use.
...
 
+### options.outputPath
+
+Type: `string`
+
+Default value: `undefined`
+
+Specify the path of the output file.  
+The destination directory must already exist.
+Also make sure the file extension corresponds with the chosen [reporter](#reporter).
```
Comment (stefanjudis) [inline README.md]: Why that?

It doesn't matter at all, does it?
I mean I just ran the `json` reporter and wrote to `.txt` - it's not a problem at all as far as I can see?

File: tasks/jsinspect.js [hunk 3]
```diff
@@ -61,8 +61,41 @@ module.exports = function(grunt) {
...
+      return function() {
+        process.stdout.write = oldWrite;
+      };
+    }
+
+    var unhook;
+    var stdoutFS;
+
+    if (typeof options.outputPath === 'string') {
+      stdoutFS = require('fs').createWriteStream(options.outputPath, {encoding: 'utf8', flags: 'w'});
```
Comment (stefanjudis) [inline tasks/jsinspect.js]: Personally I'm not a big fan of these inline `require` statements. But this is probably only my taste. :)

**Comment by stefanjudis** [comment]
@borisdiakur 

Thanks for taking the time. A had a few notes - maybe we wanna discuss these.

The actual implementation is a bit hard to understand at the beginning - but I thought about it and probably wouldn't come to a nice solution. So this might be fine. :)

Maybe we can add some comments or something? I'm not sure about it. :bowtie: 

Overall really good PR. Thanks for that. 

Maybe @paazmaya can also have a look. :)

**Comment by borisdiakur** [comment]
You're welcome. Actually I'd find it cleaner if the jsinspect module would offer the write to file part. That way we wouldn't have to use such a hack. Maybe we should send another PR to the other repo...

**Comment by stefanjudis** [comment]
> Actually I'd find it cleaner if the jsinspect module would offer the write to file part.

Agree on that - maybe we wanna keep this one open, until @danielstjules decides that he would take it?

**Comment by danielstjules** [comment]
If I was using npm scripts or make, I'd just pipe the output: `jsinspect . > yourfile` Is there a clean alternative to this with grunt?

I guess one improvement would be to update the reporters to accept the stream to write to, rather than assume process.stoudout.

**Comment by stefanjudis** [comment]
> If I was using npm scripts or make, I'd just pipe the output: jsinspect . > yourfile Is there a clean alternative to this with grunt?

Hah, was expecting this reply. ;)

> I guess one improvement would be to update the reporters to accept the stream to write to, rather than assume process.stoudout.

Might work yes.

What do you think about providing the reporter result with the `end` event?

https://github.com/stefanjudis/grunt-jsinspect/blob/master/tasks/jsinspect.js#L64-L66

E.g.

```
    inspector.on('end', function(result) {
      // do something with the result here

      done(taskSucceeded);
    });
```

**Comment by borisdiakur** [comment]
@danielstjules See https://github.com/danielstjules/jsinspect/issues/24 - I'm about to fork your repo...

**Comment by borisdiakur** [comment]
PR created: https://github.com/danielstjules/jsinspect/pull/25
heading out for :pizza:

**Comment by danielstjules** [comment]
I like the idea of accepting a writable stream. Makes integration tests super easy as well, since capturing stdout from mocha is a PITA :)

**Comment by borisdiakur** [comment]
PR https://github.com/danielstjules/jsinspect/pull/25 has been accepted. Thanks @danielstjules!
Now we have to wait until the next release of jsinspect in order to rework and complete this PR.

**Comment by stefanjudis** [comment]
Good job @borisdiakur! :tada:

**Comment by danielstjules** [comment]
Just pushed a new release :) Please note that I renamed the option. Thanks for your patience, and all your help!

**Comment by stefanjudis** [comment]
:+1: :tada: 

Nice example of open source collaboration! Love it!

**Comment by borisdiakur** [comment]
That was fast! Thanks again @danielstjules! Ok, I guess it's my turn now. Hang tight!

**Comment by borisdiakur** [comment]
The PR should be :ok_hand: now. @stefanjudis: Your turn.

File: test/jsinspect_test.js (hunk 1 above)
Comment (danielstjules) [inline test/jsinspect_test.js]: Right now `fs.rmdir` wouldn't run if any of the assertions failed. I'd suggest placing it in /tmp or the os tmp dir :) https://github.com/raszi/node-tmp :)

Comment (stefanjudis) [inline test/jsinspect_test.js]: :+1:

**Comment by borisdiakur** [comment]
[tmp](https://github.com/raszi/node-tmp) is nice!

**Comment by paazmaya** [comment]
Sweet :sweet_potato: !

**Comment by stefanjudis** [comment]
okay looks good to me. :)

**Pull request merged**